Issue11 wsdl2rest integration

bfitzpat opened the pull request:
Incorporates all the latest changes to master as well as the changes proposed by @apupier for a "fat jar" version of wsdl2rest for the generator. 

Signed-off-by: Brian Fitzpatrick <bfitzpat@redhat.com>

bfitzpat commented:
To run the generator with the additional wsdl2rest functionality, now you simply type:
```
> yo camel-project --wsdl2rest
```

This will prompt you for two additional properties:
* ? URL to the input WSDL (requires a file URL such as file:/E:/eclipse-photon/eclipse-workspace/empty-spring/src/main/resources/META-INF/wsdl/Address.wsdl)
* ? Name of the output directory for generated artifacts src//main//java (this is defaulted to src/main/java and will be where the generated CXF source goes from the wsdl you specify)

Unfortunately at this point, I'm not successful in getting the wsdl2rest utility to run in this manner. I see this output locally:
```
Creating folders
Copying files
calling: java -jar C:\Users\brianf\Documents\GitHub\generator-camel-project-fork\app\wsdl2rest\target\wsdl2rest-impl-fatjar-0.1.1-SNAPSHOT.jar -Dlog4j.configuration=file:///C:/Users/brianf/Documents/GitHub/generator-camel-project-fork/app/wsdl2rest/config/logging.properties --wsdl file:/E:/eclipse-photon/eclipse-workspace/empty-spring/src/main/resources/META-INF/wsdl/Address.wsdl --out c:\Users\brianf\Documents\demo\src\main\java --camel-context c:\Users\brianf\Documents\demo\src\main\resources\META-INF\spring\camel-context-rest.xml
identical pom.xml
identical README.md
identical src\main\resources\META-INF\spring\camel-context.xml
stderr: wsdl2rest [options...]
stderr:

stderr:  --blueprint-context PATH : Path to the generated blueprint camel context file
 --camel-context PATH     : Path to the generated spring camel context file
 --jaxrs URL              : Address of the generated jaxrs endpoint
 --jaxws URL              : Address of the target jaxws endpoint
 --out PATH               : Output path for generated artefacts (required)
 --wsdl URL               : URL to the input WSDL (required)

stderr: 1
wsdl2rest did not generate artifacts successfully - please check the log file for details
```

So this hints that a few things -- our logging path is off for some reason (though it correct) and the utility is not taking the passed-in details though they are specified in the command line. 

I'm a bit stumped here, but at least we're making some progress. Thank you @apupier for the fat-jar fixes!

bfitzpat commented:
We can add some additional command-line options as well for the passed-in wsdl URL and the path for the generated Java files, but for now I thought we'd just focus on the prompt path.

bfitzpat commented:
Working on adding additional Mocha tests now...

apupier commented on test/app.js:
```diff
@@ -169,3 +169,49 @@ describe('generator-camel:app', function () {
+        });
+    });
+
+    // it('Should create the basic structure two ways', function () {
```
commented test:
is it because it is not working yet? if yes, is it possible to put some @Ignore with a reason why or a JIRA number to work on the issue.

bfitzpat replied:
It's commented and moved up to the promise of the test before it because the wsdl2rest process has not finished by the time I'm looking for the artifacts it generates. I don't think this is actually running correctly the way I have it written and am still playing with things to see how to get this to test more effectively. Open to suggestions since I'm still trying to figure out this whole "promise" asynchronous way of doing things in JS.

bfitzpat replied:
I have uncommented and moved it back because I don't believe it was being run.

bfitzpat commented:
I've asked a question on StackOverflow to see what shakes out. I'm sure it's something simple - https://stackoverflow.com/questions/53105600/how-do-i-build-a-mocha-test-that-waits-for-a-process-to-finish-first

bfitzpat commented:
Now that we have a working test I can complete the work to get additional prompts in for optional properties. (Presuming that the test also works up on Travis)

bfitzpat commented:
Looks like mocha is failing with a timeout. I will investigate upping the timeout in the travis configuration.

apupier commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "generator-camel-project",
-  "version": "0.1.2",
+  "version": "0.1.3",
```
why do we have a diff here? Just because it would need to rebase on master?

bfitzpat replied:
would you prefer that we do version changes in separate PRs? I honestly just thought that adding the initial cut of wsdl2rest support would also include a version bump

apupier replied:
I was expecting that the master branch was already updated for the "following version".
Fine to do it in this PR

apupier commented on test/app.js:
```diff
@@ -169,3 +169,37 @@ describe('generator-camel:app', function () {
+    it('Should create the basic structure two ways', function () {
+      basicProps.name = 'MyAppMock';
+      basicProps.package = 'com.generator.mock';
+      basicProps.camelVersion = '2.18.2';
```
any reasons to choose this "old" version? What about using the latest released at the moment 2.22.2? or we want to be backward-compatible?

bfitzpat replied:
Will update to 2.22.2

apupier replied:
updated in https://github.com/camel-tooling/generator-camel-project/pull/33

apupier commented on app/wsdl2rest/pom.xml:
```diff
@@ -0,0 +1,149 @@
+    <groupId>org.github.camel.tooling.wsdl2rest</groupId>
+
+    <properties>
+        <version.wsdl2rest>0.8.0.fuse-720034</version.wsdl2rest>
```
double-checking, it is the version that has been pinned for non-deletion even if initially in staging repo?

bfitzpat replied:
Yes, this is the version that is pinned

apupier commented on app/wsdl2rest/pom.xml:
```diff
@@ -0,0 +1,149 @@
+
+    <properties>
+        <version.wsdl2rest>0.8.0.fuse-720034</version.wsdl2rest>
+        <!-- <version.wsdl2rest>0.7.0</version.wsdl2rest> -->
```
this commented out line can be removed

apupier commented on app/wsdl2rest/pom.xml:
```diff
@@ -0,0 +1,149 @@
+			<plugin>
+				<groupId>org.springframework.boot</groupId>
+				<artifactId>spring-boot-maven-plugin</artifactId>
+				<version>2.0.4.RELEASE</version>
```
latest version is now 2.1.0.RELEASE, might worth upgrading to latest and greatest on initial commit

bfitzpat commented:
Due to some funky issues I hit with a merge of readme.md, I will handle updating the readme separately with https://github.com/camel-tooling/generator-camel-project/issues/32

bfitzpat commented:
* Squashed the commits to a single commit to clean up the history and reference issue #11 .
* will handle updating the readme separately with issue #32
* re springboot jar - I am avoiding this and leaving it for now. I don't see springboot being added yet to wsdl2rest as an option and we will have to visit springboot support separately with issue #18 
* re how I got the test working...

First, there was an issue with how the wsdl2restGenerate method was calling the Java jar. Thanks to some help from Evan Shortiss, he recommended returning a Promise -- with either a resolve (success) or reject (error), and then relying on that in the test using "toPromise" to wait for the promise to resolve itself before attempting to do the asserts. 

I was missing the return of the Promise from the method upstream, which meant it really didn't have anything to wait for in my test.

bfitzpat commented:
Once this particular PR is merged I will move on to https://github.com/camel-tooling/generator-camel-project/issues/12 to finish the work.

bfitzpat commented:
Cleaned up the reported messages a bit, fixed the log4j problem, and added a --debug option. This is without it.

```
$ yo camel-project --wsdl2rest
     _                             _
    / \     _ __     __ _    ___  | |__     ___
   / _ \   | '_ \   / _` |  / __| | '_ \   / _ \
  / ___ \  | |_) | | (_| | | (__  | | | | |  __/
 /_/   \_\ | .__/   \__,_|  \___| |_| |_|  \___|
           |_|
       ____                              _
     /  ___|   __ _   _ __ ___     ___  | |
    |  |      / _` | | '_ ` _ \   / _ \ | |
    |  |___  | (_| | | | | | | | |  __/ | |
     \____|   \__,_| |_| |_| |_|  \___| |_|
 -----------------------------------------------
            Camel Project Generator
                 Version: 0.1.2
 -----------------------------------------------

? Your Camel project name demo
? Your Camel version 2.22.2
? Camel DSL type (blueprint, spring, or java) blueprint
? Package name:  com.demo
? URL to the input WSDL http://www.thomas-bayer.com/axis2/services/BLZService?wsdl
? Name of the output directory for generated artifacts src/main/java
Creating folders
Copying files
Calling wsdl2rest
   create CXF artifacts for specified WSDL at src/main/java
   create src/main/resources/OSGI-INF/blueprint/blueprint.xml
   create pom.xml
   create README.md
   create src\main\resources\OSGI-INF\log4j2.properties

brianf@brianf-work MINGW64 ~/Documents/demo
```

bfitzpat commented:
This is with the --debug option

```
$ yo camel-project --wsdl2rest --debug
     _                             _
    / \     _ __     __ _    ___  | |__     ___
   / _ \   | '_ \   / _` |  / __| | '_ \   / _ \
  / ___ \  | |_) | | (_| | | (__  | | | | |  __/
 /_/   \_\ | .__/   \__,_|  \___| |_| |_|  \___|
           |_|
       ____                              _
     /  ___|   __ _   _ __ ___     ___  | |
    |  |      / _` | | '_ ` _ \   / _ \ | |
    |  |___  | (_| | | | | | | | |  __/ | |
     \____|   \__,_| |_| |_| |_|  \___| |_|
 -----------------------------------------------
            Camel Project Generator
                 Version: 0.1.2
 -----------------------------------------------

? Your Camel project name demo
? Your Camel version 2.22.2
? Camel DSL type (blueprint, spring, or java) blueprint
? Package name:  com.demo
? URL to the input WSDL http://www.thomas-bayer.com/axis2/services/BLZService?wsdl
? Name of the output directory for generated artifacts src/main/java
Creating folders
Copying files
calling: java  -Dlog4j.configuration=file:///C:/Users/brianf/Documents/GitHub/generator-camel-project-fork/app/wsdl2rest/config/logging.properties -jar C:\Users\brianf\Documents\GitHub\generator-camel-project-fork\app\wsdl2rest\target\wsdl2rest-impl-fatjar-0.1.3-SNAPSHOT.jar --wsdl http://www.thomas-bayer.com/axis2/services/BLZService?wsdl --out c:\Users\brianf\Documents\demo\src\main --blueprint-context c:\Users\brianf\Documents\demo\src\main\resources\OSGI-INF\blueprint\blueprint.xml
stdout: Retrieving document at 'http://www.thomas-bayer.com/axis2/services/BLZService?wsdl'.

stderr: Services:

stderr:         BLZService

stderr:         Port: BLZServiceSOAP11port_http
        Binding: BLZServicePortType
        PortType: BLZServicePortType

stderr:         Operations:

stderr:                 Operation: getBank

stderr:                         Input:

stderr:                         Message: getBank

stderr:                         Part: :getBank
                                Params: com/blz.thomas-bayer.GetBank {http://thomas-bayer.com/blz/}getBank

stderr:                         Output:
                        Message: getBankResponse
                        Part: :getBankResponse
                                Params: com/blz.thomas-bayer.GetBankResponse {http://thomas-bayer.com/blz/}getBankResponse
                        Faults:
        Port: BLZServiceHttpport
        Binding: BLZServicePortType
        PortType: BLZServicePortType
        Operations:
                Operation: getBank
                        Input:
                        Message: getBank
                        Part: :getBank
                                Params: com/blz.thomas-bayer.GetBank {http://thomas-bayer.com/blz/}getBank
                        Output:
                        Message: getBankResponse
                        Part: :getBankResponse
                                Params: com/blz.thomas-bayer.GetBankResponse {http://thomas-bayer.com/blz/}getBankResponse
                        Faults:

stderr:         Port: BLZServiceSOAP12port_http
        Binding: BLZServicePortType
        PortType: BLZServicePortType
        Operations:
                Operation: getBank
                        Input:
                        Message: getBank
                        Part: :getBank
                                Params: com/blz.thomas-bayer.GetBank {http://thomas-bayer.com/blz/}getBank
                        Output:
                        Message: getBankResponse
                        Part: :getBankResponse
                                Params: com/blz.thomas-bayer.GetBankResponse {http://thomas-bayer.com/blz/}getBankResponse
                        Faults:

   create CXF artifacts for specified WSDL at src/main/java
   create src/main/resources/OSGI-INF/blueprint/blueprint.xml
   create pom.xml
   create README.md
   create src\main\resources\OSGI-INF\log4j2.properties

brianf@brianf-work MINGW64 ~/Documents/demo
```

bfitzpat commented:
Also, to address some of your suggestions/requested changes:
* stderr is now hidden unless the --debug option is set, but we can change the name if we want to something more user friendly (see https://github.com/camel-tooling/generator-camel-project/pull/31#issuecomment-439491830 vs. https://github.com/camel-tooling/generator-camel-project/pull/31#issuecomment-439492066)
* the generator now uses the Rest DSL generated by wsdl2rest as the main camel configuration file in the spring or blueprint project, ignoring the one from the template
* the generator now also grabs a different pom.xml with appropriate Rest dependencies for spring and blueprint
* the generated source goes to src/main/java vs src/main/java/java, though it's a hack. I suspect the extra java directory is coming from wsdl2rest itself though I was unable to quickly tell where that might be coming from exactly
* fixed the summary at the end (see my first note in this list)
* fixed the pom.xml spacing
* fixed a bug I discovered when using an external URL vs. a file URL and added a new test

apupier commented on app/index.js:
```diff
@@ -147,11 +182,130 @@ module.exports = class extends yeoman {
+}
+
+function wsdl2restGenerate(wsdlUrl, outputDirectory, dsl, isDebug) {
+    if (dsl.indexOf('java') > 0) {
```
Eclipse is suggesting me to use include instead of "dsl.indexOf('java') > 0"
it seems more readable, what do you think?

apupier commented on app/index.js:
```diff
@@ -147,11 +182,130 @@ module.exports = class extends yeoman {
+        cmdString = cmdString + ' --camel-context ' + restContextPath;
+    }
+    if (isDebug) {
+        console.log('calling: ' + cmdString);
```
I'm wondering if it wouldn't be better to keep the same log in normal and debug mode, and just add more information when in debug mode. it would help to compare normal and debug logs, what do you think?

```
console.log('Calling wsdl2rest');
 if (isDebug) {
        console.log('command used: ' + cmdString);
}
```

apupier commented on app/index.js:
```diff
@@ -147,11 +182,130 @@ module.exports = class extends yeoman {
+
+    var wsdl2restdir = path.join(__dirname, 'wsdl2rest');
+    var targetDir = path.join(wsdl2restdir, 'target');
+    var jar = path.join(targetDir, 'wsdl2rest-impl-fatjar-0.1.3-SNAPSHOT.jar');
```
would it be possible to avoid hardcoding the version here?
I fear that developers will forget to update it often.

bfitzpat replied:
Not sure how I would do that.  Maybe try a wildcard? wsdl2rest-impl-fatjar-*.jar ?

lhein replied:
just do a lookup in the folder with a filenamefilter?

apupier commented on app/index.js:
```diff
@@ -147,11 +182,130 @@ module.exports = class extends yeoman {
-            );
+            var skipFile = false;
+            if (showWsdl2Rest) {
+                skipFile = (testFileForWsdl2RestSkip(this.files[i]) == true);
```
is it a limitation of javascript to have to compare for boolean before storing into a boolean?
why not directly:
skipFile = testFileForWsdl2RestSkip(this.files[i]);

bfitzpat replied:
No, it's not a limitation -- was a design decision. If we're in "wsdl2rest" mode, I want it to check the list to see if we should skip the file. If not, I only want to skip the one file. But by default I don't want to accidentally skip a file.

apupier replied:
so there is a difference between
`skipFile = (testFileForWsdl2RestSkip(this.files[i]) == true);`
and
`skipFile = testFileForWsdl2RestSkip(this.files[i]);`
?

why it would skip the file by default with the second line?

bfitzpat replied:
Adjusted the code to try and make it a bit clearer and make the non-wsdl2rest case consistent.

apupier commented on app/index.js:
```diff
@@ -147,11 +182,130 @@ module.exports = class extends yeoman {
+
+function wsdl2restGenerate(wsdlUrl, outputDirectory, dsl, isDebug) {
+    if (dsl.indexOf('java') > 0) {
+        console.log(`Generating Rest DSL from SOAP for a Camel Java DSL is currently unsupported.`);
```
I was able to generate the REST DSl choosing java DSl:

```
yo camel-project --wsdl2rest
     _                             _
    / \     _ __     __ _    ___  | |__     ___
   / _ \   | '_ \   / _` |  / __| | '_ \   / _ \
  / ___ \  | |_) | | (_| | | (__  | | | | |  __/
 /_/   \_\ | .__/   \__,_|  \___| |_| |_|  \___|
           |_|
       ____                              _
     /  ___|   __ _   _ __ ___     ___  | |
    |  |      / _` | | '_ ` _ \   / _ \ | |
    |  |___  | (_| | | | | | | | |  __/ | |
     \____|   \__,_| |_| |_| |_|  \___| |_|
 -----------------------------------------------
            Camel Project Generator
                 Version: 0.1.2
 -----------------------------------------------

? Your Camel project name (testCommandLine)
? Your Camel project name testCommandLine
? Your Camel version (2.22.2)
? Your Camel version 2.22.2
? Camel DSL type (blueprint, spring, or java) (java)
? Camel DSL type (blueprint, spring, or java) java
? Package name:  (com.testCommandLine)
? Package name:  koko.pl
? URL to the input WSDL (C:\Users\Aurelien Pupier\Desktop\Address.wsdl)
? URL to the input WSDL C:\Users\Aurelien Pupier\Desktop\Address.wsdl
? Name of the output directory for generated artifacts (src/main/java)
? Name of the output directory for generated artifacts src/main/java
Creating folders
Copying files
Calling wsdl2rest
   create CXF artifacts for specified WSDL at src/main/java
   create src/main/resources/META-INF/spring/camel-context.xml
   create README.md
   create src\main\java\koko\pl\routes\CamelRoute.java
   create src\main\java\koko\pl\routes\Launcher.java
```

bfitzpat replied:
Interesting. You should have seen an error -- I'll check it out.

bfitzpat replied:
Seems my validation method disappeared during one of my redo's - I'll work on getting that back in.

apupier commented on app/index.js:
```diff
@@ -147,11 +182,130 @@ module.exports = class extends yeoman {
+
+    // build the java command with classpath, class name, and the passed parameters
+    var cmdString = 'java ';
+    cmdString = cmdString + ' -Dlog4j.configuration=' + logUrl;
```
can't we write cmdString += ' -Dlog4j.configuration=' + logUrl; in Javascript?

bfitzpat replied:
I don't know - but I'm not a fan of that syntax regardless. I think it's clearer when written out.

apupier commented on app/index.js:
```diff
@@ -147,11 +182,130 @@ module.exports = class extends yeoman {
+    restContextPath = path.join(process.cwd(), rawContextPath);
+
+    // build the java command with classpath, class name, and the passed parameters
+    var cmdString = 'java ';
```
what do you think to write it:
```
var cmdString = 'java ' +
                           ' -Dlog4j.configuration=' + logUrl +
                           ' --wsdl ' + wsdlFileUrl +
                           ' --out ' + outPath;
```
see https://google.github.io/styleguide/javascriptguide.xml?showone=Multiline_string_literals#Multiline_string_literals

bfitzpat replied:
Yes, that's a bit cleaner.

apupier commented on app/index.js:
```diff
@@ -147,11 +182,130 @@ module.exports = class extends yeoman {
+        rawContextPath = "src/main/resources/OSGI-INF/blueprint/blueprint.xml";
+    } else {
+        rawContextPath = "src/main/resources/META-INF/spring/camel-context.xml";
+    };
```
the ; is not needed here

bfitzpat commented:
Added the --jaxrsURL and --jaxwsURL prompts to the mix to try and work through the weird camel context generation issue. Also updated the helloworld.wsdl to be more namespace appropriate -- continuing to investigate 

Have it a little closer
```
<to uri="cxf://http://localhost:3000/helloworldservice/?serviceClass=wsdl.com/wsdl/HelloService.examples.www.HelloPortType&amp;defaultOperationName=sayHello"/>
```

But the issue seems to be with the fully qualified class name (that is generated correctly) that's not coming across to populate this part of the velocity template - https://github.com/jboss-fuse/wsdl2rest/blob/cb42273e7ad7ebe573d03bc5e41983316fbbd6b3/impl/src/main/resources/templates/wsdl2rest-spring-context.vm#L57

bfitzpat commented:
I think it's the targetnamespace specified in the wsdl... trying that approach.

bfitzpat commented:
That was the trick. Now you can go through the steps in https://github.com/camel-tooling/generator-camel-project/pull/31#issuecomment-442282769 without modifying the generated camel context file to see it run end to end with a working soap service.

@apupier Let me know if this works for you?

bfitzpat commented:
Note: Don't bother testing the command line options yet - I'm still fine tuning. There are some weird issues with how info is being passed from the command-line arguments down to the wsdl2rest call, so I'm trying to work on that.

bfitzpat commented:
The only thing remaining is to test the generated Rest service to see if it actually runs. I'll look into doing that tomorrow. Since we create a soap service and we're creating all the right bits, we may be able to call into maven to get it to build and run (mvn install; mvn camel;run) and then see if we can ping the REST get operation.

apupier commented:
the normal Java generation is not working anymore.

```
C:\git\generator-camel-project\normalJava>yo camel-project --debug
     _                             _
    / \     _ __     __ _    ___  | |__     ___
   / _ \   | '_ \   / _` |  / __| | '_ \   / _ \
  / ___ \  | |_) | | (_| | | (__  | | | | |  __/
 /_/   \_\ | .__/   \__,_|  \___| |_| |_|  \___|
           |_|
       ____                              _
     /  ___|   __ _   _ __ ___     ___  | |
    |  |      / _` | | '_ ` _ \   / _ \ | |
    |  |___  | (_| | | | | | | | |  __/ | |
     \____|   \__,_| |_| |_| |_|  \___| |_|
 -----------------------------------------------
            Camel Project Generator
                 Version: 0.1.2
 -----------------------------------------------

? Your Camel project name normalJava
? Your Camel version 2.22.2
? Camel DSL type (blueprint, spring, or java)
>> When using wsdl2rest, the Camel DSL must be either 'spring' or 'blueprint'.
```

apupier commented on test/app.js:
```diff
@@ -22,15 +22,42 @@ var helpers = require('yeoman-test');
+    });
+
+    it('utilities findWsdl2RestJar should succeed in finding wsdl2rest jar', function () {
+      // test runs on its own but fails in the larger test suite - still figuring that out
```
is it still true? there is no error on PR CI job

bfitzpat replied:
Not true - will remove PR

apupier commented:
with next fix can you rebase on master please?
it will avoid to see modifications related to this commit for isntance https://github.com/camel-tooling/generator-camel-project/commit/20b969c3311c7dfc01e7f5ad733bb558ebed1b22 and simpligy code review

apupier commented on app/index.js:
```diff
@@ -69,7 +88,17 @@ module.exports = class extends yeoman {
+
+        var showW2EPrompts = true;
+        if (!showPrompts && showWsdl2Rest) {
+            showW2EPrompts = !(utils.isNotNull(this.options.wsdl)) &&
```
IS it possible to use something like utils.isNull || utils.isNull ? it will avoid the double negation "!" "isNot"

bfitzpat replied:
Yup!

apupier commented:
when trying to access http://localhost:8081/jaxrs/hello/%22Aurelien%22

I have this error:
```
org.apache.cxf.interceptor.Fault: Could not send Message.
	at org.apache.cxf.interceptor.MessageSenderInterceptor$MessageSenderEndingInterceptor.handleMessage(MessageSenderInterceptor.java:67)
	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
	at org.apache.cxf.endpoint.ClientImpl.doInvoke(ClientImpl.java:537)
	at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:439)
	at org.apache.camel.component.cxf.CxfProducer.process(CxfProducer.java:133)
	at org.apache.camel.processor.SendProcessor.process(SendProcessor.java:148)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:548)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:138)
	at org.apache.camel.processor.Pipeline.process(Pipeline.java:101)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201)
	at org.apache.camel.component.direct.DirectBlockingProducer.process(DirectBlockingProducer.java:53)
	at org.apache.camel.processor.SendProcessor.process(SendProcessor.java:148)
	at org.apache.camel.processor.RedeliveryErrorHandler.process(RedeliveryErrorHandler.java:548)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201)
	at org.apache.camel.processor.CamelInternalProcessor.process(CamelInternalProcessor.java:201)
	at org.apache.camel.component.jetty.CamelContinuationServlet.doService(CamelContinuationServlet.java:215)
	at org.apache.camel.http.common.CamelServlet.service(CamelServlet.java:78)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:841)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:564)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:317)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:110)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.Invocable.invokePreferred(Invocable.java:128)
	at org.eclipse.jetty.util.thread.Invocable$InvocableExecutor.invoke(Invocable.java:222)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:294)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:199)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:673)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:591)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: ConnectException invoking http://localhost:8080/somepath: Connection refused: connect
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.mapException(HTTPConduit.java:1402)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1386)
	at org.apache.cxf.transport.AbstractConduit.close(AbstractConduit.java:56)
	at org.apache.cxf.transport.http.HTTPConduit.close(HTTPConduit.java:673)
	at org.apache.cxf.interceptor.MessageSenderInterceptor$MessageSenderEndingInterceptor.handleMessage(MessageSenderInterceptor.java:63)
	... 41 more
Caused by: java.net.ConnectException: Connection refused: connect
	at java.net.DualStackPlainSocketImpl.waitForConnect(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:85)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at sun.net.NetworkClient.doConnect(NetworkClient.java:175)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:463)
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:558)
	at sun.net.www.http.HttpClient.<init>(HttpClient.java:242)
	at sun.net.www.http.HttpClient.New(HttpClient.java:339)
	at sun.net.www.http.HttpClient.New(HttpClient.java:357)
	at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1220)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1199)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050)
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:984)
	at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1334)
	at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1309)
	at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream.setupWrappedStream(URLConnectionHTTPConduit.java:275)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleHeadersTrustCaching(HTTPConduit.java:1345)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.onFirstWrite(HTTPConduit.java:1306)
	at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream.onFirstWrite(URLConnectionHTTPConduit.java:307)
	at org.apache.cxf.io.AbstractWrappedOutputStream.write(AbstractWrappedOutputStream.java:47)
	at org.apache.cxf.io.AbstractThresholdOutputStream.write(AbstractThresholdOutputStream.java:69)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1358)
	... 44 more
```

I thought about a port occupied, tried a 'netstat -an' but the port 8080 is not used.

bfitzpat commented:
Yes, you should change the JAXWS endpoint in the prompts to http://localhost:3000/helloworldservice/ as I suggested here:
```
$ yo camel-project --wsdl2rest
     _                             _
    / \     _ __     __ _    ___  | |__     ___
   / _ \   | '_ \   / _` |  / __| | '_ \   / _ \
  / ___ \  | |_) | | (_| | | (__  | | | | |  __/
 /_/   \_\ | .__/   \__,_|  \___| |_| |_|  \___|
           |_|
       ____                              _
     /  ___|   __ _   _ __ ___     ___  | |
    |  |      / _` | | '_ ` _ \   / _ \ | |
    |  |___  | (_| | | | | | | | |  __/ | |
     \____|   \__,_| |_| |_| |_|  \___| |_|
 -----------------------------------------------
            Camel Project Generator
                 Version: 0.1.2
 -----------------------------------------------

? Your Camel project name demo
? Your Camel version 2.22.2
? Camel DSL type (blueprint, spring, or java) spring
? Package name:  com.demo
? URL to the input WSDL http://localhost:3000/helloworldservice?wsdl
? Name of the output directory for generated artifacts src/main/java
? Address of the generated jaxrs endpoint http://localhost:8081/jaxrs
? Address of the target jaxws endpoint http://localhost:3000/helloworldservice/
Creating folders
Copying files
Calling wsdl2rest
   create CXF artifacts for specified WSDL at src/main/java
   create src/main/resources/META-INF/spring/camel-context.xml
   create pom.xml
```

bfitzpat commented on app/index.js:
```diff
@@ -109,23 +140,73 @@ module.exports = class extends yeoman {
+            }, prompts);
+        }
+
+        if (showPrompts || showW2EPrompts) {
```
I will change to "showW2RPrompts"

bfitzpat commented:
> the normal Java generation is not working anymore.

Try without "--debug" -- that's a wsdl2rest only feature at this point. And perhaps we should change the option to be "--w2rdebug" or something more descriptive?

bfitzpat commented:
As for why it "works" when you test it vs. when you run through the prompts in the test it avoids the validation I think.

bfitzpat commented:
And I have created a new issue for the problem with not picking up the soap address from the wsdl - https://github.com/jboss-fuse/wsdl2rest/issues/72

bfitzpat commented:
Addressed the java DSL problem, fixed some of the minor issues you found @apupier, and rebased on master. I've asked for some help from a couple of people for the new test that builds and runs the camel service in maven and then will do a http get, just like the manual test I had you do, but that shouldn't hold us up since we have a manual workaround in the meantime. 

This PR provides a functionally complete feature with some additional work to be done as we get more comfortable with JavaScript, but I think we're in good shape.

EricWittmann commented on test/app_maven.js:
```diff
@@ -0,0 +1,63 @@
+        console.log('trying to run mvn camel:run in ' + mvnpath);
+        try {
+            const spawn = require('child_process').spawn;
+            const child = spawn('C:/apache-maven-3.5.2/bin/mvn --version');
```
Try this:

```
const child = spawn('mvn', [ '--version' ]);
```

bfitzpat commented:
For now I have "skipped" the test that is not working to test that we can build, run, and access the running rest service. We will continue to work on figuring out the issues with the test, but it should not block us from getting this functionality out to our users. We can then focus on improving that functionality as we go forward.

bfitzpat commented:
The Mocha/Yeoman framework we are using does not handle validating each prompt as we go through them, but I can likely write a test for the potential regression. Will see what I can do about that and Blueprint.

bfitzpat commented:
@apupier Added a test for the camel dsl validation (with and without wsdl2rest). And dropped a note to Freeman and Claus about the Blueprint issue

bfitzpat commented:
I got some feedback from Freeman about running the generated Blueprint example in Standalone mode. It seems some changes would need to be made to make that happen. See this new issue I added to wsdl2rest - https://github.com/jboss-fuse/wsdl2rest/issues/73

@apupier If you want to test the generated project locally, you can follow those steps to get it to run standalone and then do the same "http://localhost:8081/jaxrs/hello/%22Aurelien%22" test on it to pass through Camel to the SOAP service and back again. 

So this will require some upstream changes to make it work for Blueprint in a testable capacity. What's there now should work for a deployable case (deploying to a Karaf instance), but won't work for a standalone test. 

As a result, I think we can call this done for Spring, but state that Blueprint will require some additional work.

apupier commented:
> As a result, I think we can call this done for Spring, but state that Blueprint will require some additional work.

agree. i think if we put a comment in readme for now, it will be enough to move forward with this PR on this point.

apupier commented:
there is a test which is not finishing https://travis-ci.org/camel-tooling/generator-camel-project/builds/462869962#L4053-L4058

```
- Should create the basic structure and CXF files for spring with an internal running WSDL url and attempt to build the project, run it, and test that the REST service is accessible
Stopped web service on http://localhost:3000/helloworldservice
  17 passing (15s)
  1 pending
```

it is blocked when locally launching "npm test", I guess it says "pending" on CI because there is a timeout set in the package;json

bfitzpat commented:
Very strange. Pending is the same as a skip and it never gets stuck here. Are you using "npm test" instead of just calling "mocha"?

bfitzpat commented:
Very strange. It never gets stuck here. And it's not getting stuck on CI. Maybe do "npm link" locally to clean up any lingering code from older builds and try "npm test" again?

bfitzpat commented:
Are you sure you have rebased it? I only have
```
  16 passing (15s)
  1 pending
```

apupier commented:
> Are you sure you have rebased it? I only have
> 
> ```
>   16 passing (15s)
>   1 pending
> ```

17 is from the CI

![image](https://user-images.githubusercontent.com/1105127/49460242-2a150600-f7f1-11e8-954c-88cdc4bdd439.png)

apupier commented:
> Very strange. Pending is the same as a skip and it never gets stuck here. Are you using "npm test" instead of just calling "mocha"?

yes, using "npm test", I would expect it to work :(

bfitzpat commented:
My bad on the number of tests - I forgot I added a new test on my other machine and didn't sync this one.  Now I get:
```
  17 passing (15s)
  1 pending
```

That said, I still have no clue why it's failing on your box. Is it possible you have a Java process that's hanging somewhere?

apupier commented:
> That said, I still have no clue why it's failing on your box. Is it possible you have a Java process that's hanging somewhere?

there are not. I don't understand why another java process would hang the test execution.

apupier commented on test/app_maven.js:
```diff
@@ -0,0 +1,68 @@
+            // Resolve or reject based on the result of the camel run.
+            execCamelRun().then(() => resolve()).catch((e) => reject(e));
+        }).catch((e) => reject(e));
+        // console.log('Running mvn install');
```
commented code to remove

bfitzpat replied:
Removed file, which isn't needed any longer when I removed the skipped test

bfitzpat commented:
> there are not. I don't understand why another java process would hang the test execution.

I have had instances where the Terminal inside (or outside) VSCode has caused a hung Java process on my Windows box. Just wanted to suggest that as a possibility. Thanks for checking.

bfitzpat commented:
@apupier I have split up the test into three pieces -- generator_tests, wsdl2rest_tests, and utils_tests -- and removed the maven/chai testing that was skipped.

Can you give it a shot again?
